src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientServiceExtensions.cs

```diff
@@ -0,0 +1,83 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Net.Http;
+using Grpc.AspNetCore.Server.GrpcClient;
+using Grpc.Core;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Http;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class GrpcClientServiceExtensions
+    {
+        public static IHttpClientBuilder AddGrpcClient<TClient>(this IServiceCollection services, Action<GrpcClientOptions<TClient>> configureClient) where TClient : ClientBase<TClient>
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configureClient == null)
+            {
+                throw new ArgumentNullException(nameof(configureClient));
+            }
+
+            services.TryAdd(ServiceDescriptor.Transient(typeof(ITypedHttpClientFactory<TClient>), typeof(GrpcHttpClientFactory<TClient>)));
+            services.TryAdd(ServiceDescriptor.Transient(typeof(GrpcHttpClientFactory<TClient>.Cache), typeof(GrpcHttpClientFactory<TClient>.Cache)));
+
+            services.Configure(configureClient);
+
+            // Accessing the client options here allows the HttpContextAccessor to be added as needed
+            // Is there a way to get the value from services.Configure now?
+            var clientOptions = new GrpcClientOptions<TClient>();
+            configureClient(clientOptions);
+
+            // HttpContextAccessor has performance overhead. Only add it when required
+            if (RequireHttpContextAccessor(clientOptions))
+            {
+                services.AddHttpContextAccessor();
```

Comment:
I have heard that `AddHttpContextAccessor` introduces overhead to each request so I made it conditionally added depending on whether the client will use values from the current request. Is this a true fact?

Reply:
It is a true fact. I want to know more about what this is for.

Reply:
Yes there are also thread safety concerns. This is about cancellation propagation.

Reply:
Microservices tend to call other microservices. A gRPC client using the current requests cancellation token will allow its calls to be cancelled when the current request is cancelled.

Cancellation token, deadline and context propagation token are values that could be set from the current request.